File: JavaScript/leetcode/shift-2d-grid/solution.js
/**
 * @param {number[][]} grid
 * @param {number} k
 * @return {number[][]}
 */
var shiftGrid = function(grid, k) {
    const m = grid.length;
    const n = m > 0 ? grid[0].length : 0;
    const total = m * n;
    if (total === 0) return grid;
    k %= total;
    const flat = [];
    for (let i = 0; i < m; i++) for (let j = 0; j < n; j++) flat.push(grid[i][j]);
    const res = Array.from({length: m}, () => new Array(n).fill(0));
    for (let idx = 0; idx < total; idx++) {
        const np = (idx + k) % total;
        res[Math.floor(np / n)][np % n] = flat[idx];
    }
    return res;
};

File: Scala/leetcode/shift-2d-grid/solution.scala
object Solution {
    def shiftGrid(grid: Array[Array[Int]], k: Int): Array[Array[Int]] = {
        val m = grid.length; val n = grid(0).length; val total = m * n; val kk = k % total
        val res = Array.ofDim[Int](m, n)
        for (i <- 0 until total) { val ni = (i + kk) % total; res(ni / n)(ni % n) = grid(i / n)(i % n) }
        res
    }
}
